### PR TITLE
net: fix no sessions error if O hostname doesn't match

### PR DESCRIPTION
**Pull request by cyberj0g**
**What does this pull request do? Explain your changes. (required)**
Issue #2485 
Particularly nasty bug with observable issue occurring in unrelated session management logic, but straightforward resolution. The root cause is OrchestratorPool returning empty OrchestratorLocalInfo, if O's `-serviceAddr` doesn't match B's `-orchAddr` precisely (e.g. **localhost** won't match **127.0.0.1**). It might have caused silent O selection issues in any case when `-orchAddr` or `-serviceAddr` were specified.

**Specific updates (required)**

- Add host name resolution function, with support for `localhost`

**How did you test each of these updates (required)**

**Does this pull request close any open issues?**

Issue #2485 

**Checklist:**

- [ ] Read the [contribution guide](./doc/contributing.md)
- [ ] `make` runs successfully
- [ ] All tests in `./test.sh` pass
- [ ] README and other documentation updated
- [ ] [Pending changelog](./CHANGELOG_PENDING.md) updated

### CODE REVIEW

File: discovery/discovery.go [hunk 1]
```diff
@@ -55,12 +59,65 @@ func (o *orchestratorPool) GetInfos() []common.OrchestratorLocalInfo {
...
+	queryUrl, _ := url.Parse(uri)
+	resolvedUrls := ResolveURL(queryUrl)
 	for _, info := range o.infos {
-		if info.URL.String() == uri {
-			res = info
-			break
+		oResolvedUrls := ResolveURL(info.URL)
+		for _, oUrl := range oResolvedUrls {
+			for _, queryUrl := range resolvedUrls {
+				if oUrl.String() == queryUrl.String() {
```
Comment (leszko) [inline discovery/discovery.go]:
```suggestion
			for _, qUrl := range resolvedUrls {
				if oUrl.String() == qUrl.String() {
```

Nit: you already have the `queryUrl` defined.

File: discovery/discovery.go [hunk 2]
```diff
@@ -55,12 +59,65 @@ func (o *orchestratorPool) GetInfos() []common.OrchestratorLocalInfo {
...
 func (o *orchestratorPool) GetInfo(uri string) common.OrchestratorLocalInfo {
 	var res common.OrchestratorLocalInfo
+	queryUrl, _ := url.Parse(uri)
+	resolvedUrls := ResolveURL(queryUrl)
 	for _, info := range o.infos {
-		if info.URL.String() == uri {
-			res = info
-			break
+		oResolvedUrls := ResolveURL(info.URL)
+		for _, oUrl := range oResolvedUrls {
```
Comment (leszko) [inline discovery/discovery.go]:
```suggestion
		for _, oUrl := range ResolveURL(info.URL) {
```

Nit: inline

File: discovery/discovery.go [hunk 3]
```diff
@@ -55,12 +59,65 @@ func (o *orchestratorPool) GetInfos() []common.OrchestratorLocalInfo {
 	return o.infos
 }
 
+func ResolveURL(urlToResolve *url.URL) []*url.URL {
+	resolvedUrls := make([]*url.URL, 0, 0)
+	// check if there's a port in host, for ipv6 it's after ]
```
Comment (leszko) [inline discovery/discovery.go]: Isn't `urlToResolve` already parsed to `host` and `port`. I think that `urlToResolve.Host` should contain host (without port) and `urlToResolve.Port()` should contain the port.

If that's not the case, I guess we should just use `url.ParseRequestURI()` instead of this custom logic below.

File: discovery/discovery.go [hunk 4]
```diff
@@ -55,12 +59,65 @@ func (o *orchestratorPool) GetInfos() []common.OrchestratorLocalInfo {
 	return o.infos
 }
 
+func ResolveURL(urlToResolve *url.URL) []*url.URL {
+	resolvedUrls := make([]*url.URL, 0, 0)
```
Comment (leszko) [inline discovery/discovery.go]:
```suggestion
	var resolvedUrls []*url.URL
```

Nit.

File: discovery/discovery.go [hunk 5]
```diff
@@ -55,12 +59,65 @@ func (o *orchestratorPool) GetInfos() []common.OrchestratorLocalInfo {
...
+		var addrs []string
+		var err error
+		if strings.ToLower(host) == "localhost" {
+			// ipv6 addr may be abbreviated or not
+			addrs = []string{"127.0.0.1", "::1", "0000:0000:0000:0000:0000:0000:0000:0001"}
+		} else {
+			addrs, err = hostLookupFn(host)
+		}
+		if err != nil {
+			clog.Errorf(context.TODO(), "Couldn't resolve domain for URL %s, check local DNS settings. Discovery may not work properly.", urlToResolve.String())
```
Comment (leszko) [inline discovery/discovery.go]:
```suggestion
			glog.Errorf("Couldn't resolve domain for URL %s, check local DNS settings", urlToResolve.String())
```
Nit

File: discovery/discovery.go [hunk 6]
```diff
@@ -55,12 +59,65 @@ func (o *orchestratorPool) GetInfos() []common.OrchestratorLocalInfo {
 	return o.infos
 }
 
+func ResolveURL(urlToResolve *url.URL) []*url.URL {
```
Comment (leszko) [inline discovery/discovery.go:62]: Should we always add the input `urlToResolve` to `resolvedUrls`, just in case we missed something in the logic below, but the addresses are already equal? That would be safer in terms of backward-compatibility.

File: server/broadcast.go [hunk 7]
```diff
@@ -722,6 +722,10 @@ func selectOrchestrator(ctx context.Context, n *core.LivepeerNode, params *core.
 			bcastOS = bcastOS.OS().NewSession(pfx)
 		}
 
+		oinfo := n.OrchestratorPool.GetInfo(tinfo.Transcoder)
```
Comment (leszko) [inline server/broadcast.go:725]: Is this change related to your PR? This seems like a candidate for a separate PR.

**Review by leszko — CHANGES_REQUESTED** [review]
@cyberj0g Thanks for the PR. Added some comments. But in general it's a very necessary change!

File: discovery/discovery.go (hunk 1 above)
Comment (cyberj0g) [reply discovery/discovery.go]: Fixed.

File: discovery/discovery.go (hunk 3 above)
Comment (cyberj0g) [reply discovery/discovery.go]: Thanks for suggestion, replaced with library functions.

File: discovery/discovery.go (hunk 5 above)
Comment (cyberj0g) [reply discovery/discovery.go]: Fixed.

File: server/broadcast.go (hunk 7 above)
Comment (cyberj0g) [reply server/broadcast.go:725]: Yes, that what was making this bug hard to track down. `GetInfo` silently returns uninitialized structure in case there's no URL match. Ideally, we need to refactor it to return error, but it will require changes to the interface, probably should be a separate PR.

File: discovery/discovery.go (hunk 2 above)
Comment (cyberj0g) [reply discovery/discovery.go]: Fixed.

File: discovery/discovery.go (hunk 4 above)
Comment (cyberj0g) [reply discovery/discovery.go]: Fixed.

File: discovery/discovery.go (hunk 6 above)
Comment (cyberj0g) [reply discovery/discovery.go:62]: Added backward compatibility shortcut to `GetInfo()`.

**Review by leszko — APPROVED** [review]

File: server/broadcast.go (hunk 7 above)
Comment (RiccardoBiosas) [inline server/broadcast.go:725]: is it the reason why the bug was occurring only when establishing a session for the transcoding?
Because when a broadcaster first submits an orchestrator's domain, it is parsed in a way that is supposed to match the format of the `serviceAddr` - except apparently for the localhost/127.0.0.1 edge-case. In fact in the case of https://github.com/livepeer/go-livepeer/issues/2485 not prefixing the O domain `solarfarm.multiorch.com` with `https` is not a problem right away because the [`parseOrchAddrs` function](https://github.com/livepeer/go-livepeer/blob/master/cmd/livepeer/starter/starter.go#L1191-L1209) takes care of reformatting it - if so, rather than creating a new function (`ResolveURL`) can't we reuse the same util used in `starter.go` (with the appropriate changes to handle the ipv6/localhost edge-cases)?
